js/pro/poloniex.js

```diff
@@ -0,0 +1,886 @@
+        const filledAmount = this.safeString (order, 'filledAmount'); // TODO? filledQuantity
+        let trades = undefined;
+        if (!Precise.stringEq (filledAmount, '0')) {
+            trades = [];
+            const trade = this.parseWsOrderTrade (order);
+            trades.push (trade);
+        }
+        return this.safeOrder ({
+            'info': order,
+            'symbol': this.safeSymbol (marketId, market),
+            'id': id,
+            'clientOrderId': clientOrderId,
+            'timestamp': timestamp,
+            'datetime': this.iso8601 (timestamp),
+            'lastTradeTimestamp': undefined,
+            'type': this.safeString (order, 'type'),
+            'timeInForce': undefined,
+            'postOnly': undefined,
+            'side': this.safeString (order, 'side'),
+            'price': this.safeString (order, 'price'),
+            'stopPrice': undefined,
+            'triggerPrice': undefined,
+            'amount': this.safeString (order, 'orderAmount'),
+            'cost': undefined,
+            'average': undefined,
+            'filled': filledAmount,
+            'remaining': this.safeString (order, 'remaining_size'),
+            'status': undefined, // TODO?: eventType, state
+            'fee': {
+                'rate': undefined,
+                'cost': this.safeString (order, 'tradeFee'),
+                'currency': this.safeString (order, 'feeCurrency'),
+            },
+            'trades': trades,
+        });
+    }
+
+    handleTicker (client, message) {
+        //
+        //    {
+        //        channel: 'ticker',
+        //        data: [
+        //            {
+        //                symbol: 'BTC_USDT',
+        //                startTime: 1677280800000,
+        //                open: '23154.32',
+        //                high: '23212.21',
+        //                low: '22761.01',
+        //                close: '23148.86',
+        //                quantity: '105.179566',
+        //                amount: '2423161.17436702',
+        //                tradeCount: 17582,
+        //                dailyChange: '-0.0002',
+        //                markPrice: '23151.09',
+        //                closeTime: 1677367197924,
+        //                ts: 1677367251090
+        //            }
+        //        ]
+        //    }
+        //
+        const data = this.safeValue (message, 'data');
+        if (data !== undefined) {
+            for (let i = 0; i < data.length; i++) {
+                const item = data[i];
+                const marketId = this.safeString (item, 'symbol');
+                if (marketId !== undefined) {
+                    const ticker = this.parseTicker (item);
+                    const symbol = ticker['symbol'];
+                    this.tickers[symbol] = ticker;
+                    const messageHash = 'ticker:' + marketId;
+                    client.resolve (ticker, messageHash);
```

Comment:
I think we're missing the case when the user calls `watchTickers` without any args (`poloniex.watchTickers()`). When that happens we should resolve a future without any symbol attached `messageHash = 'ticker'`

Reply:
I don't know how to deal with this within `handleTicker` because the response is an array which will will always have a marketId, and I don't know how to distinguish between them

Reply:
you can resolve for both, by having two client resolve. The one you have to manage the watchTicker listening to a specific ticker and adding another on line`client.resolve(this.tickers, 'ticker')` on line 695

Reply:
I still don't think I understand this

Reply:
so it's almost working now, but `watchTickers` is only returning the first object in the array of the response object, I don't really understand why the others aren't being returned because theres a loop on line 713, does the `resolve` call terminate the rest of the loop?

```
      close: '0.002119',
      quantity: '161.6',
      amount: '0.3357085',
      tradeCount: 97,
      dailyChange: '0.0043',
      markPrice: '0.002134',
      closeTime: 1679097181022,
      ts: 1679097189285
    },
    {
      symbol: 'DOGE_BTC',
      startTime: 1679010780000,
      open: '0.000002917',
      high: '0.000002917',
      low: '0.000002765',
      close: '0.000002796',
      quantity: '288208',
      amount: '0.811046409',
      tradeCount: 197,
      dailyChange: '-0.0415',
      markPrice: '0.000002795',
      closeTime: 1679097181012,
      ts: 1679097189285
    },
    ... 541 more items
  ]
}
{
  id: 'BTS_BTC',
  symbol: 'BTS/BTC',
  timestamp: 1679097189285,
  datetime: '2023-03-17T23:53:09.285Z',
  high: 4.692e-7,
  low: 4.301e-7,
  bid: undefined,
  bidVolume: undefined,
  ask: undefined,
  askVolume: undefined,
  vwap: 4.42599602461e-7,
  open: 4.451e-7,
  close: 4.301e-7,
  last: 4.301e-7,
  previousClose: undefined,
  change: -1.5e-8,
  percentage: -3.3700292069197935,
  average: 4.376e-7,
  baseVolume: 26161,
  quoteVolume: 0.0115788482,
  info: {
    symbol: 'BTS_BTC',
    startTime: 1679010780000,
    open: '0.0000004451',
    high: '0.0000004692',
    low: '0.0000004301',
    close: '0.0000004301',
    quantity: '26161',
    amount: '0.0115788482',
    tradeCount: 21,
    dailyChange: '-0.0337',
    markPrice: '0.000000438',
    closeTime: 1679097181021,
    ts: 1679097189285
  }
}
```

Reply:
The messageHash of `ticker:all` gets deleted from `this.futures` inside `resolve`, so it's no longer there for the 2nd item in data

Reply:
yep, that's because after calling client.resolve the client does two things, 

1) it resolves the promise stored in the messageHash under client.futures[messageHash]
2) it deletes the messageHash index with `delete client.futures[messageHash]`

this is done on purpose for async purposes, so that two calls to exchange.watch return the same promise. In the case that two resolves are done on the same messageHash without any new promises being created by new calls to exchange.watchX the second one will be ignored (there is a check inside of client.resolve)

Reply:
> yep, that's because after calling client.resolve the client does two things,
> 
>     1. it resolves the promise stored in the messageHash under client.futures[messageHash]
> 
>     2. it deletes the messageHash index with `delete client.futures[messageHash]`
> 
> 
> this is done on purpose for async purposes, so that two calls to exchange.watch return the same promise. In the case that two resolves are done on the same messageHash without any new promises being created by new calls to exchange.watchX the second one will be ignored (there is a check inside of client.resolve)

so whats the solution? Do I need to assign `client.futures[messageHash]` every loop iteration?

Reply:
> yep, that's because after calling client.resolve the client does two things,
> 
>     1. it resolves the promise stored in the messageHash under client.futures[messageHash]
> 
>     2. it deletes the messageHash index with `delete client.futures[messageHash]`
> 
> 
> this is done on purpose for async purposes, so that two calls to exchange.watch return the same promise. In the case that two resolves are done on the same messageHash without any new promises being created by new calls to exchange.watchX the second one will be ignored (there is a check inside of client.resolve)

so whats the solution? do I need to assign `client.futures[messageHash]` to a new promise every loop iteration?

Reply:
> so whats the solution? do I need to assign `client.futures[messageHash]` to a new promise every loop iteration?

I tried doing this, which didn't really seem like the right solution, but `watchTickers` is still only returning the first item of the array

Reply:
You don't need to create a future here, the future is created is created and returned by the watch function.
In the `handleTicker` you want to resolve the future that has already been created when calling `watch`. And they are two cases:
- client.resolve (this.tickers[symbol], 'ticker:' + marketId) -> Called once for every ticker -> returns the value to watchTicker 
- client.resolve (this.tickers, `ticker:all`) -> Called once for all tickets -> returns the value to watchTickers